api/api/admin/media_report.py

```diff
@@ -106,6 +193,165 @@ def pending_reports_links(self, obj):
+
+        if request.method == "POST":
+            expiration = self.lock_manager.add_locks(
+                request.user.get_username(), object_id
+            )
+            return JsonResponse(
+                data={"expiration": expiration},
+                status=503 if expiration == 0 else 200,
+            )
+
+        return redirect(f"admin:api_{self.media_type}_change", object_id)
+
+    #################
+    # Moderate view #
+    #################
+
+    def moderate_view(self, request, object_id):
+        """
+        Create a decision for the media object and associate selected
+        reports complaining about the media with this decision.
+        """
+
+        if request.method == "POST":
+            media_obj = self.get_object(request, object_id)
+
+            form = MediaDecisionForm(request.POST, media_type=self.media_type)
+            if form.is_valid():
+                decision_class, through_class = {
+                    "image": (ImageDecision, ImageDecisionThrough),
+                    "audio": (AudioDecision, AudioDecisionThrough),
+                }[self.media_type]
+
+                action = form.cleaned_data["action"]
+                notes = form.cleaned_data["notes"]
+                moderator = request.user
+                decision = decision_class.objects.create(
+                    action=action,
+                    notes=notes,
+                    moderator=moderator,
+                )
+                logger.info(
+                    "Decision created",
+                    decision=decision.id,
+                    action=action,
+                    notes=notes,
+                    moderator=moderator.username,
+                )
+
+                through = through_class.objects.create(
+                    decision=decision,
+                    media_obj=media_obj,
+                )
+                logger.info(
+                    "Through model created",
+                    through=through.id,
+                    decision=decision.id,
+                    media_obj=media_obj.id,
+                )
+
+                for report in form.cleaned_data["reports"]:
+                    report.decision = decision
+                    report.save()
+                    logger.info(
+                        "Report updated",
+                        report=report.id,
+                        decision=decision.id,
+                    )
+            else:
+                logger.warning(
+                    "Form is invalid",
+                    **form.cleaned_data,
+                    errors=form.errors,
+                )
+
+        return redirect(f"admin:api_{self.media_type}_change", object_id)
+
+    #############
+    # Overrides #
+    #############
+
+    # TODO: This construct breaks down if a decision is associated with
+    #   a media item that does not have any reports. Such an item cannot
+    #   be reached at the URL ``admin/api/{media_type}/{id}/change/``.
```

Comment:
This endpoint only exists for media items in the queryset, which only includes items that have at least one report. For items that are not in the queryset, this URL doesn't work.

Reply:
I mean the view exists but when `self.get_object(request, object_id)` is called, `get_object` returns `None` because the ID we're requesting is not in the queryset.

Reply:
I see. That isn't really tenable. Can it include all works, not just ones with reports? Moderators can sort by the number of pending reports, or we can add a filter action that only shows works in need of moderation (i.e., show the "moderation queue").

Reply:
We can consider that, yes, I'm not sure if that should be in this PR though (it's pretty huge as it is currently)!